src/runtime/virtcontainers/qemu.go

```diff
@@ -1020,47 +1049,138 @@ func (q *qemu) togglePauseSandbox(pause bool) error {
 }
 
 func (q *qemu) qmpSetup() error {
-	q.qmpMonitorCh.Lock()
-	defer q.qmpMonitorCh.Unlock()
+	return q.startQMPInstance(&q.qmpMonitorCh, "monitor-qmp", true)
+}
 
-	if q.qmpMonitorCh.qmp != nil {
+func (q *qemu) startQMPInstance(qmpChan *qmpChannel, subsystem string, enableEvents bool) error {
+	qmpChan.Lock()
+	defer qmpChan.Unlock()
+
+	if qmpChan.qmp != nil {
 		return nil
 	}
 
-	cfg := govmmQemu.QMPConfig{Logger: newQMPLogger()}
+	var events chan govmmQemu.QMPEvent
+
+	if enableEvents {
+		events = make(chan govmmQemu.QMPEvent)
+		go q.loopQMPEvent(events)
+	}
+
+	// It's safe to pass a nil EventCh
+	cfg := govmmQemu.QMPConfig{
+		Logger: qmpLogger{
+			logger: virtLog.WithField("subsystem", subsystem),
+		},
+		EventCh: events,
+	}
 
 	// Auto-closed by QMPStart().
 	disconnectCh := make(chan struct{})
 
-	qmp, _, err := govmmQemu.QMPStart(q.qmpMonitorCh.ctx, q.qmpMonitorCh.path, cfg, disconnectCh)
+	qmp, _, err := govmmQemu.QMPStart(qmpChan.ctx, qmpChan.path, cfg, disconnectCh)
 	if err != nil {
 		q.Logger().WithError(err).Error("Failed to connect to QEMU instance")
 		return err
 	}
 
-	err = qmp.ExecuteQMPCapabilities(q.qmpMonitorCh.ctx)
+	err = qmp.ExecuteQMPCapabilities(qmpChan.ctx)
 	if err != nil {
 		qmp.Shutdown()
 		q.Logger().WithError(err).Error(qmpCapErrMsg)
 		return err
 	}
-	q.qmpMonitorCh.qmp = qmp
-	q.qmpMonitorCh.disconn = disconnectCh
+	qmpChan.qmp = qmp
+	qmpChan.disconn = disconnectCh
+
+	return nil
+}
+
+func (q *qemu) loopQMPEvent(event chan govmmQemu.QMPEvent) {
+	for {
+		select {
+		case e, open := <-event:
+			if !open {
+				q.Logger().Infof("QMP event channel closed")
+				return
+			}
+			q.Logger().Debugf("got new QMP event %+v", e)
+			if e.Name == "GUEST_PANICKED" {
+				go q.handleGuestPanic()
+			}
+		}
+	}
+}
+
+func (q *qemu) handleGuestPanic() {
+	if q.config.GuestMemoryDumpEnabled {
+		q.dumpGuestMemory()
+	}
+	// TODO: how to notify the upper level sandbox to handle the error
+	// to do a fast fail(shutdown or others).
+}
+
+func (q *qemu) dumpGuestMemory() error {
+	q.Logger().Info("try to dump guest memory")
+	dumpSavePath := q.config.GuestMemoryDumpPath
+	if dumpSavePath == "" {
+		dumpSavePath = dumpSavePath
+	}
 
+	if fi, err := os.Stat(dumpSavePath); err != nil {
+		if os.IsNotExist(err) {
+			os.MkdirAll(dumpSavePath, 0755)
+		} else {
+			return err
+		}
+	} else if !fi.IsDir() {
+		return fmt.Errorf("Not a directory: %s", dumpSavePath)
+	}
+
+	protocol := fmt.Sprintf("file:%s/vmcore-%s.%s", dumpSavePath, q.id, q.config.GuestMemoryDumpFormat)
```

Comment:
- Shouldn't this be `file://%s/vmcore-%s.%s`?
- Maybe the name should include something to indicate it's a kata-triggered / Qemu / guest kernel dump rather than a host kernel dump? Particularly since `/var/crash/` is primarily for host dumps and the kernel versions (host and guest) are bound to be different. Maybe we should also be writing a `.metadata` file that includes deatils of the version of qemu + guest kernel file used alongside this dump file to make analysis easier?

Reply:
@jodh-intel I think the content of `metadata` may something like `kata-runtime kata-env`, so it's better reusing some code of `kata-env`.

But `kata-env` is in `pkg/cli/` package and have many private `func`s/`var`s, so in this PR I will only append information from `HypervisorConfig` that `qemu` holding.

The dump file and meta information will be stored under `<DUMP_PATH>/<SANDBOX_ID>`,  and includes:
- dump file, named by `format("vmcore-%s.%s", time.Now().Format("20060102150405.999"), format)`
- hypervisor config
- hypervisor version

Reply:
Thanks @liubin. Sounds good to me.